pkg/osquery/runtime/osquery_instance_history/history.go

```diff
@@ -0,0 +1,139 @@
+package osquery_instance_history
+
+import (
+	"errors"
+	"sync"
+	"time"
+)
+
+const maxInstances = 10
+
+var currentHistory = &history{}
+
+type history struct {
+	sync.Mutex
+	instances []*Instance
+}
+
+type NoCurrentInstanceError struct{}
+
+func (c NoCurrentInstanceError) Error() string {
+	return "no osquery instance is currently set"
+}
+
+// InstanceStarted adds a new instance to the osquery instance history
+func InstanceStarted() error {
+	currentHistory.Lock()
+	defer currentHistory.Unlock()
+
+	return currentHistory.nextInstance()
+}
+
+// InstanceConnected sets the connect time and instance id of the current osquery instance
+func InstanceConnected(querier Querier) error {
+	currentHistory.Lock()
+	defer currentHistory.Unlock()
+
+	return currentHistory.setCurrentInstanceConnected(querier)
+}
+
+// InstanceExited sets the exit time and appends provided error (if any) to current osquery instance
+func InstanceExited(exitError error) error {
+	currentHistory.Lock()
+	defer currentHistory.Unlock()
+
+	return currentHistory.setCurrentInstanceExited(exitError)
+}
+
+// GetHistory returns the last 10 instances of osquery started / restarted by launcher, each start / restart cycle is an entry
+func GetHistory() []Instance {
+	currentHistory.Lock()
+	defer currentHistory.Unlock()
+
+	results := make([]Instance, len(currentHistory.instances))
+	for i, v := range currentHistory.instances {
+		results[i] = *v
+	}
+	return results
```

Comment:
It's a question of access, I expect the values of these histories to be immutable after they are initially set (except for maybe errors).  By de-referencing we're creating a copy and returning that, any consumer of this package will not have the ability to edit the history, forcing them to use exported functions.

This does in fact come at the cost of extra allocations, but since we are maxing at 10, I feel it's worth it for proper encapsulation.